ledger/internal/evalprefetcher_test.go

```diff
@@ -0,0 +1,675 @@
+			txnGroups[i][k].SignedTxn = transactions.SignedTxn{
+				Txn: transactions.Transaction{
+					Type: protocol.PaymentTx,
+					Header: transactions.Header{
+						Sender: acctAddr(1),
+					},
+					PaymentTxnFields: transactions.PaymentTxnFields{
+						Receiver:         acctAddr(addr),
+						CloseRemainderTo: acctAddr(addr + 1),
+					},
+				},
+			}
+			addr += 2
+		}
+	}
+	preloadedTxnGroupsCh := prefetchAccounts(context.Background(), ledger, rnd, txnGroups, feeSinkAddr, config.Consensus[protocol.ConsensusCurrentVersion])
+	groupsCount := 0
+	addressCount := 0
+	uniqueAccounts := make(map[basics.Address]bool)
+	for k := range preloadedTxnGroupsCh {
+		groupsCount++
+		addressCount += len(k.accounts)
+		for _, acct := range k.accounts {
+			uniqueAccounts[*acct.address] = true
+		}
+	}
+	require.Equal(t, len(txnGroups), groupsCount)
+	// the +1 below is for the fee sink address.
+	require.Equal(t, len(txnGroups)*16*3+1, addressCount)
+	require.Equal(t, len(txnGroups)*16*2+1, len(uniqueAccounts))
+}
+
+func BenchmarkPrefetcherApps(b *testing.B) {
+	acctAddrPtr := func(i int) (o *basics.Address) {
+		o = new(basics.Address)
+		o[0] = byte(i)
+		o[1] = byte(i >> 8)
+		o[2] = byte(i >> 16)
+		return
+	}
+	acctAddr := func(i int) (o basics.Address) {
+		t := *acctAddrPtr(i)
+		copy(o[:], t[:])
+		return
+	}
+
+	txnGroupLen := 16
+	groups := make([][]transactions.SignedTxnWithAD, 1+b.N/txnGroupLen)
+	for grpIdx := range groups {
+		groups[grpIdx] = make([]transactions.SignedTxnWithAD, txnGroupLen)
+		for txnIdx := range groups[grpIdx] {
+			groups[grpIdx][txnIdx].SignedTxn = transactions.SignedTxn{
+				Txn: transactions.Transaction{
+					Type: protocol.ApplicationCallTx,
+					Header: transactions.Header{
+						Sender: acctAddr(grpIdx + txnIdx),
+					},
+					ApplicationCallTxnFields: transactions.ApplicationCallTxnFields{
+						ApplicationID: 10,
+						Accounts: []basics.Address{
+							acctAddr(grpIdx + txnIdx + 1),
+							acctAddr(grpIdx + txnIdx + 1),
+						},
+						ForeignApps: []basics.AppIndex{
+							2001,
+							2002,
+						},
+						ForeignAssets: []basics.AssetIndex{
+							1001,
+						},
+					},
+				},
+			}
+		}
+	}
+	rnd := basics.Round(5)
+	var feeSinkAddr = basics.Address{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
+	var ledger = &prefetcherTestLedger{
+		round:    rnd,
+		balances: make(map[basics.Address]ledgercore.AccountData),
+		creators: make(map[basics.CreatableIndex]basics.Address),
+	}
+	ledger.balances[acctAddr(1)] = ledgercore.AccountData{
+		AccountBaseData: ledgercore.AccountBaseData{MicroAlgos: basics.MicroAlgos{Raw: 100000000}},
+	}
+
+	b.ResetTimer()
+	preloadedTxnGroupsCh := prefetchAccounts(context.Background(), ledger, rnd, groups, feeSinkAddr, config.Consensus[protocol.ConsensusCurrentVersion])
+	for k := range preloadedTxnGroupsCh {
```

Comment:
please store them and compare to the input (at least size)

Reply:
it's not just the size.. the slices provided in the channel are the *exact* same slices in the original groups - which is why I used indexing before..

Reply:
I understand. And it needs to tested to ensure all groups are propagated.

Reply:
covered in TestEvaluatorPrefetcherQueueExpansion, but I'll update.

Reply:
oh, sorry, I misscrolled down to the benchmark. TestEvaluatorPrefetcherQueueExpansion is good enogth